src/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -559,10 +560,19 @@ public static string UnescapeDataString(string stringToUnescape)
 
                     UnescapeMode unescapeMode = UnescapeMode.Unescape | UnescapeMode.UnescapeAll;
                     position = 0;
-                    char[] dest = new char[stringToUnescape.Length];
-                    dest = UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, dest, ref position,
+                    ValueStringBuilder pooledArray = new ValueStringBuilder(stringToUnescape.Length);
+                    UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, ref pooledArray, ref position,
                         c_DummyChar, c_DummyChar, c_DummyChar, unescapeMode, null, false);
-                    return new string(dest, 0, position);
+
+                    if (stringToUnescape.Length == position && pooledArray.AsSpan().SequenceEqual(new ReadOnlySpan<char>(pStr, position)))
+                    {
+                        pooledArray.Dispose();
+                        return stringToUnescape;
+                    }
+
+                    string result = pooledArray.AsSpan(0, position).ToString();
+                    pooledArray.Dispose();
+                    return result;
```

Comment:
Nit: seems like you could simplify this slightly:
```C#
Span<char> resultSpan = pooledarray.AsSpan(0, position);
string resultString = resultSpan.SequenceEqual(new ReadOnlySpan<char>(pStr, position)) ? stringToUnescape : resultSpan.ToString();
pooledArray.Dispose();
return result;
```

Reply:
Seems to be a valid point in general, but what about stringToUnescape.Length == position check which is removed from your proposal? It allows to avoid invoking SequenceEqual when lengths are different which is supposedly slightly faster.

Reply:
That check is performed by SequenceEqual, which is also aggressively inlined.

Reply:
I simplified the code, but had to keep stringToUnescape.Length == position check because we need to make sure that position points to the end of the original string. SequenceEqual doesn't check it.

Reply:
> I simplified the code, but had to keep stringToUnescape.Length == position check because we need to make sure that position points to the end of the original string. SequenceEqual doesn't check it.

Can you elaborate?  Why is `resultSpan.SequenceEqual(stringToUnescape)` insufficient?

Reply:
My comment refers to the original version of your comment where you suggested to use `resultSpan.SequenceEqual(new ReadOnlySpan<char>(**pStr, position**)).` This expression doesn't take into account the `stringToUnescape.Length`.  The latest suggested version `.SequenceEqual(stringToUnescape)` is probably sufficient.

Reply:
> My comment refers to the original version of your comment

Yes, GitHub doesn't yet have an integrated compiler for code comments ;)